File: Python/SinglyLinkedList/singly_linked_list.py
# create linked list with no built-in methods
# Create Node to add to list
class Node:
  def __init__(self, value):
    self.value = value
    self.next = None

# Create a Singly Linked List
class SinglyLinkedList:
  def __init__(self, value):
    self.head = Node(value)
  # add node to back of list
  def addToBack(self, value):
    runner = self.head
    while runner:
      if runner.next == None:
        runner.next = Node(value)
        break
      runner = runner.next
    return self
  # add node to front of list
  def addToFront(self, value):
    currentHead = self.head
    newHead = Node(value)
    self.head = Node(value)
    self.head.next = currentHead
    return self
  # remove node from back of list
  def removeBack(self):
    runner = self.head
    while runner:
      if runner.next.next == None:
        runner.next = None
      runner = runner.next
    return self
  # remove node from beginning of list
  def removeFront(self):
    self.head = self.head.next
    return self
  # insert value at specific index
  def insertValue(self, value, idx):
    runner = self.head
    i = 0
    while runner:
      if idx == 0:
        self.addToFront(value)
      else:
        if idx == i + 1:
          newNode = Node(value)
          next = runner.next
          runner.next = newNode
          newNode.next = next
          break
      i+=1
      runner = runner.next
    return self
  # remove value at specific index
  def removeIndex(self, idx):
    if idx < 0:
      return self
    else:
      runner = self.head
      i = 0;
      while runner:
        if idx == 0:
          self.removeFront()
          break
        else:
          if idx - 1 == i and runner.next != None:
            if runner.next.next != None:
              runner.next = runner.next.next
            else:
              runner.next = None
        i+=1
        runner = runner.next
      return self
  # remove first value given
  def removeValue(self, value):
    runner = self.head
    if runner.value == value:
      self.head = self.head.next
    while runner:
      if runner.next != None:
        if runner.next.value == value and runner.next.next != None:
          runner.next = runner.next.next
          break
        if runner.next.value == value and runner.next.next == None:
          runner.next = None
          break
      # if runner.next.value == val:
        # pass
      runner = runner.next
    return self
  # display nodes in list
  def printValues(self):
    runner = self.head
    i = 0
    arr = []
    while runner:
      node = runner
      arr.append(node.value)
      str = 'This is the current node {}, containing value {}, the next node is {},'.format(i, runner.value, runner.next)
      print(str)
      runner = node.next
      i += 1
    print(arr)
    return self
myList = SinglyLinkedList(10).addToBack(11).addToBack(12).addToBack(13).addToBack(14).printValues().addToFront(9).removeBack().printValues().removeFront().printValues().insertValue(100, 2).printValues().removeIndex(4).addToBack(16).printValues().removeValue(16).printValues()

File: JavaScript/SinglyLinkedList/singlyLinkedList.js
// create linked list with no built -in methods

// Create Node to add to list
class Node {
  constructor(value) {
    this.value = value;
    this.next = null;
  }
}

// Create SLL
class SinglyLinkedList {
  // init method
  constructor(value) {
    this.head = new Node(value);
  }
  // add node to end of list
  addToBack = value => {
    let runner = this.head;
    while(runner) {
      if (runner.next === null){
        runner.next = new Node(value);
        return this;
      }
      runner = runner.next;
    }
    return this;
  }
  // add node to front of list
  addToFront = value => {
    let runner = this.head;
    this.head = new Node(value);
    this.head.next = runner;
    return this;
  }
  // display values 
  printValues = _=> {
    let runner = this.head;
    const list = [];
    while (runner) {
      list.push(runner.value);
      console.log(`current node is ${runner.value} next node is ${runner.next}`);
      runner = runner.next;
    }
    console.log(list);
    return this;
  }
  // remove last node
  removeBack = _=> {
    let runner = this.head;
    while(runner.next) {
      if (runner.next.next === null) {
        runner.next = null;
        return this;
      }
      runner = runner.next;
    }
    return this;
  }
  // remove first node
  removeFront = _=> {
    if (this.head.next === null) {
      return this;
    }
    this.head = this.head.next;
    return this;
  }
  // insert value at index
  insertValue = (value, idx) => {
    let runner = this.head;
    let i = 0;
    while(runner && idx !== undefined) {
      if (idx == 0) {
        this.addToFront(value);
        return this;
      }
      if (i + 1 === idx) {
        let newNode = new Node(value);
        let next = runner.next;
        runner.next = newNode;
        newNode.next = next;
        return this;
      }
      runner = runner.next;
      i++;
    }
    return this;
  }
  // remove index
  removeIndex = idx => {
    let runner = this.head;
    if (idx < 0) {
      return this;
    } 
    if (idx === 0) {
      this.removeFront();
      return this;
    }
    let i = 0;
    while(runner){
      if (i + 1 === idx) {
        runner.next = runner.next.next;
        return this;
      }
      runner = runner.next;
      i++;
    }
    return this;
  }
  // remove first instance of value
  removeValue = value => {
    let runner = this.head;
    if (runner.value === value) {
      this.removeFront();
      return this;
    }
    while(runner.next) {
      if (runner.next.value === value) {
        runner.next = runner.next.next;
        return this;
      }      
      runner = runner.next;
    }
    return this;
  }
}


exports.SinglyLinkedList = SinglyLinkedList;
exports.Node = Node;